### PR TITLE
Bug fix for CPU quantization

### PR DESCRIPTION
**Pull request by vectozavr**
**🔥Quantization on CPU**
I was able to run the quantization on CPU with tiny LLama and red_pajama dataset.
Here are some small remark I made to fix some bugs including:

* When we dont have GPU devices we should add only one device: cpu in the array of devices
* We should use pin_memory only when GPU is available
* We should delete the following assert:

```
assert train_inps[i].is_pinned() and train_outs[i].is_pinned()
```

because it does not make sense since we do not use pinned memory without GPU.

### CODE REVIEW

**Review by justheuristic — APPROVED** [review]

File: main.py [hunk 1]
```diff
@@ -102,12 +103,13 @@ def get_inps(
...
     dtype = next(iter(model.parameters())).dtype
     nsamples_per_device = (len(data) - 1) // len(devices) + 1
+
     inps = [
         torch.zeros(
             (min(nsamples_per_device, len(data) - i * nsamples_per_device), model_seqlen, model.config.hidden_size),
             dtype=dtype,
             device=devices[i] if not offload_activations else "cpu",
-            pin_memory=offload_activations,
+            pin_memory=offload_activations if torch.cuda.is_available() else False,
```
Comment (galqiwi) [inline main.py:112]: 1) I think that the better way to check if we use cpu training will be to check if `devices == [torch.device('cpu')]`. This way user with gpus will be able to use cpu training with `--devices cpu` flag.
2) Let's use `A and B` instead of `A if B else False` statement (if you don't have any good reason to do otherwise).
3) I see a lot of code regarding offload_activations. Does cpu training work without this flag? Looks like it should.

File: main.py [hunk 2]
```diff
@@ -76,6 +76,7 @@ def get_inps(
     """
     print("catching layer inputs from data", flush=True)
     layers = get_layers(model)
+
```
Comment (galqiwi) [inline main.py:79]: Let's avoid changing lines that are not related to the PR. This messes up `git blame`. Thanks

Comment (galqiwi) [reply main.py:79]: > Auto-formatting does that on its own sometimes.
It does. Still, It's better to fix it.

Comment (vectozavr) [reply main.py]: Fixed ✅

File: main.py [hunk 3]
```diff
@@ -107,7 +107,7 @@ def get_inps(
             (min(nsamples_per_device, len(data) - i * nsamples_per_device), model_seqlen, model.config.hidden_size),
             dtype=dtype,
             device=devices[i] if not offload_activations else "cpu",
-            pin_memory=offload_activations,
+            pin_memory=offload_activations and devices != [torch.device('cpu')],
```
Comment (galqiwi) [inline main.py:110]: Since training works without --offload_activations flag, we don't need this change anymore

Comment (vectozavr) [reply main.py:110]: Without this additional condition we will cause the exception if we run the program having only CPU and giving --offload_activations flag
But anyway, if you think this is better I can change it

File: src/finetune.py [hunk 4]
```diff
@@ -49,7 +49,6 @@ def finetune_groupwise(
             )
         else:
             assert train_inps[i].device == train_outs[i].device == torch.device("cpu")
-            assert train_inps[i].is_pinned() and train_outs[i].is_pinned()
```
[the remark this answers is not in the document]
Comment (galqiwi) [reply src/finetune.py:52]: P.S. Sorry for late comment, forgot to publish it

File: main.py (hunk 3 above)
Comment (galqiwi) [reply main.py]: Thanks. My point is that offload_activations on cpu does not make sense, and shouldn't be supported

File: src/finetune.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vectozavr) [reply src/finetune.py]: Fixed ✅

**Comment by galqiwi** [comment]
LGTM, thanks for the PR!
